## Script to determinate the target genes 
## for a transcript form narrowPeak file generated by MaCS2.

## Author: Mercedes Brenes & Elena Cort√©s
## Contact: mebreal@gmail.com - elena.cortes.t@gmail.com
## Date: November 2019

## Installing Chipseeker annotation package of the organism.

## Create a function which turns narrowPeak into a parameter.

args <- commandArgs(trailingOnly = TRUE)

input.file.name <- args[[1]]
promoter.length <- as.numeric(args[[2]])
output.file.name <- args[[3]]

library(ChIPseeker)
library(TxDb.Athaliana.BioMart.plantsmart28)
txdb <- TxDb.Athaliana.BioMart.plantsmart28

library(clusterProfiler)
library(org.At.tair.db)
library("pathview")


## Leer el fichero de picos
read_peaks <- readPeakFile(peakfile = "peaks_1_peaks.narrowPeak",header=FALSE)
head(read_peaks)


## Defining the area around TSS considered as promoter
promoter <- getPromoters(TxDb=txdb, 
                         upstream=-1000, 
                         downstream=1000)


## number of genes AT genome

genes <- as.data.frame(genes(txdb))
genes_names <- genes$gene_id
length(genes_names)

## Peak annotation
peakAnno <- annotatePeak(peak = read_peaks, 
                              tssRegion=c(-1000, 1000),
                              TxDb=txdb)


## Binding sites in specific regions of the genome
plotAnnoPie(peakAnno)
vennpie(peakAnno)

## Distribution of genomic loci relative to TSS
plotDistToTSS(peakAnno,
              title="Distribution of genomic loci relative to TSS",
              ylab = "Genomic Loci (%) (5' -> 3')")

## Converting annotation into data frame
annot_data_frame <- as.data.frame(peakAnno)
target_genes <- annot_data_frame$geneId[annot_data_frame$annotation == "Promoter"]
length(target_genes)

write(x = target_genes,file = "my_target_genes.txt")


## GO terms enrichment

## reading target genes
gene.set <- read.table(file = "my_target_genes.txt", header = F, as.is = T) [[1]]
length(gene.set)

ego <- enrichGO(gene= gene.set, OrgDb = org.At.tair.db, ont = "BP", pvalueCutoff = 0.05, qvalueCutoff= 0.01, universe = genes_names, keyType = "TAIR")
ego.res <- as.data.frame(ego)
head(ego.res)

barplot(ego, showCategory = 13)
emapplot(ego, showCategory = 13)


## KEGG enrichmente

kk <- enrichKEGG(gene = gene.set, organism= "ath04712", universe = genes_names)
kk
kk.res <- as.data.frame(kk)
head(kk.res)

my.universe <- rep(0, length(genes_names))
names(my.universe) <- genes_names
my.universe[gene.set] <- 1
my.universe

pathway <- pathview(gene.data = my.universe, pathway.id = "ath04712", species = "ath", gene.idtype = "TAIR")



